aspnetcore/data/ef-rp/complex-data-model.md

```diff
@@ -0,0 +1,643 @@
+
+The Instructor-to-Courses many-to-many relationship:
+
+* Requires a join table that must be represented by an entity set.
+* Is a pure join table (table without payload).
+
+It's common to name a join entity `EntityName1EntityName2`. For example, the Instructor-to-Courses join table using this pattern is `CourseInstructor`. However, we recommend using a name that describes the relationship.
+
+Data models start out simple and grow. No-payload joins (PJTs) frequently evolve to include payload. By starting with a descriptive entity name, the name doesn't need to change when the join table changes. Ideally, the join entity would have its own natural (possibly single word) name in the business domain. For example, Books and Customers could be linked with a join entity called Ratings. For the Instructor-to-Courses many-to-many relationship, `CourseAssignment` is preferred over `CourseInstructor`.
+
+### Composite key
+
+FKs are not nullable. The two FKs in `CourseAssignment` (`InstructorID` and `CourseID`) together uniquely identify each row of the `CourseAssignment` table. `CourseAssignment` doesn't require a dedicated PK. The `InstructorID` and `CourseID` properties function as a composite PK. The only way to specify composite PKs to EF is with the *fluent API*. The next section shows how to configure the composite PK.
+
+The composite key ensures:
+
+* Multiple rows are allowed for one course.
+* Multiple rows are allowed for one instructor.
+* Multiple rows for the same instructor and course is not allowed.
+
+The `Enrollment` join entity defines its own PK, so duplicates of this sort are possible. To prevent such duplicates:
+
+* Add a unique index on the FK fields, or
+* Configure `Enrollment` with a primary composite key similar to `CourseAssignment`. For more information, see [Indexes](https://docs.microsoft.com/ef/core/modeling/indexes).
+
+## Update the DB context
+
+Add the following highlighted code to *Data/SchoolContext.cs*:
+
+[!code-csharp[Main](intro/samples/cu/Data/SchoolContext.cs?name=snippet_BeforeInheritance&highlight=15-18,25-31)]
+
+The preceding code adds the new entities and configures the `CourseAssignment` entity's composite PK.
+
+## Fluent API alternative to attributes
+
+The `OnModelCreating` method in the preceding code uses the *fluent API* to configure EF behavior. The API is called "fluent" because it's often used by stringing a series of method calls together into a single statement. The [following code](https://docs.microsoft.com/ef/core/modeling/#methods-of-configuration) is an example of the fluent API:
+
+```csharp
+protected override void OnModelCreating(ModelBuilder modelBuilder)
+{
+    modelBuilder.Entity<Blog>()
+        .Property(b => b.Url)
+        .IsRequired();
+}
+```
+
+In this tutorial, the fluent API is used only for DB mapping that can't be done with attributes. However, the fluent API can specify most of the formatting, validation, and mapping rules that can be done with attributes.
```

Comment:
I think it would be good to call out that some of the attributes used previously were for validation, some were for EF configuration, and some applied to both. The EF configuration parts can be changed to use fluent API instead, but the attributes used for validation cannot be replaced by EF fluent calls. EF does not do validation beyond what is needed to make coherent calls to the database.